### ISSUE TITLE
Add support for multiple versions to middleware

### ISSUE THREAD

**Issue by mcmire** [interagent/committee#74]
I love the idea of having middleware in place to check request and response data, but it doesn't appear that they support multiple versions. The middleware take a `schema` option, but if you have multiple versions this schema won't be known ahead of time and you have to choose one. (I started adding directories to represent versions to `docs/schema` -- is this proper?) What do y'all think about this?

**Comment by brandur** [comment]
@mcmire One option here is to use something like Rack's `map` to prefix a request, and then pass the prefix into Committee like so:

``` ruby
map "/v1" do
  use Committee::Middleware::RequestValidation, prefix: "/v1"
  run MyApp
end
```

It's also possible (but not built in) to write a small Rack helper to do the same for a version that comes in via another channel like an `Accept` header. In our code, we have something that looks a little like this:

``` ruby
version "3", "edge" do
  use Committee::Middleware::RequestValidation, prefix: "/v1"
  run MyApp
end
```

Adding the versions to `docs/schema` should also be a pretty good solution (although it may be more difficult to do certain things like build different versions independently I guess).

**Comment by mcmire** [comment]
I see. I don't use a URL prefix for the version at all, it solely comes in through Accept. It appears that if I reconfigured my schema to know about a URL prefix then I could use the `:prefix` option as you described above, otherwise I'm a little SOL.

Either way, don't you have to pass a `:schema` option to the middleware?

**Comment by brandur** [comment]
> I see. I don't use a URL prefix for the version at all, it solely comes in through Accept. It appears that if I reconfigured my schema to know about a URL prefix then I could use the :prefix option as you described above, otherwise I'm a little SOL.

Ah! Actually, you should still be okay, but you'll have to use something that's not `map`. We have a custom helper that looks a little something like this and depends on [sinatra-router](https://github.com/brandur/sinatra-router):

``` ruby
version "3", "edge" do
  ...
end

    # yield to a builder block in which all defined apps will only respond for
    # the given version
    def version(*versions, &block)
      condition = lambda { |env|
        versions.include?(env["HTTP_X_HEROKU_API_VERSION"])
      }
      with_conditions(condition, &block)
    end
```

> Either way, don't you have to pass a :schema option to the middleware?

Oh yes! Sorry, I missed that in my example above.

Do you think that would work for you?

**Comment by mcmire** [comment]
Oh I see, so I'd just have:

``` ruby
version "v1" do
  use Committee::Middleware::RequestValidation, schema: lookup_schema_for("v1")
  run MyApp
end

version "v2" do
  use Committee::Middleware::RequestValidation, schema: lookup_schema_for("v2")
  run MyApp
end
```

Is that the idea?

**Comment by brandur** [comment]
@mcmire Exactly right. It takes a bit of Rack-fu to get the whole thing right though.

**Comment by mcmire** [comment]
Gotcha. Well, I'm using Rails in this case, not Sinatra, so it does make things a little more difficult. What do you think about adding an option to the middleware? Something like:

``` ruby
use Committee::Middleware::RequestValidation,
  schema: MyApp::ApiSchema.new(version: 1).read,
  if: ApiVersionContraint.new(version: 1)
```

**Comment by brandur** [comment]
@mcmire Could do! It seems like something like that is a fairly general routing strategy that may not belong in any particular middleware though. I think you're right though in that it currently seems to be too difficult to add something will probably be a pretty standard requirement, and you might be pretty close to the design of a mechanism that we should provide to make things a little easier.

Do you have a `config.ru`? I'm no Rails expert, but I think that you should be able to get access to your full set of middleware there, and you could potentially add in a router around Committee.

**Comment by mcmire** [comment]
Yeah, Rails apps do come with a `config.ru`, which you are free to modify. In practice you don't have to, though, since you can also add middleware through this `config.middleware` mechanism I mentioned above. So you have options -- if the solution ends up being designed more for Rack apps then I can probably finagle it to fit more within Rails-land, since there's some commonality there.

**Comment by kyprifog** [comment]
This seems related, in rails I'm having need to using a different schema for my middleware validation if certain headers are provided:

```
if request.headers['SCHEMA-CUSTOM']
    config.middleware.use Committee::Middleware::ResponseValidation, schema: schema_custom
else
   config.middleware.use Committee::Middleware::ResponseValidation, schema: schema
end
```

How could I achieve this using your recommendations?

**Comment by mcmire** [comment]
Since I posted this, I ended up re-implementing RequestValidation as a `before_action`. It worked better for my needs. I think that might be the best option here.

**Comment by brandur** [comment]
+1 to what @mcmire said in that a more custom solution might get you the best flexibility here.

Another option that might be worth investigating if there's enough demand would be to allow the `schema` parameter to be specified as a lambda that takes `env` as a parameter. This would allow a schema to be selected dynamically for any given request.
